### PR TITLE
Introduce Column in Protocol Buffers

### PR DESCRIPTION
**Pull request by brfrn169**
Please take a look!

### CODE REVIEW

File: rpc/src/main/proto/scalardb.proto [hunk 1]
```diff
@@ -29,8 +31,22 @@ message Value {
...
+  oneof value {
+    bool boolean_value = 2;
+    int32 int_value = 3;
+    int64 bigint_value = 4;
+    float float_value = 5;
+    double double_value = 6;
+    string text_value = 7;
+    bytes blob_value = 8;
+  }
+}
```
Comment (brfrn169) [inline rpc/src/main/proto/scalardb.proto:45]: Added `Column` to ProtoBuf that corresponds to  `com.scalar.db.io.Column<T>`.

File: rpc/src/main/proto/scalardb.proto [hunk 2]
```diff
@@ -10,7 +10,9 @@ import "google/protobuf/empty.proto";
 
 // Common
 
+// Value is deprecated as of release 3.6.0. Will be removed in release 5.0.0
 message Value {
```
Comment (brfrn169) [inline rpc/src/main/proto/scalardb.proto:14]: `Value` in Protobuf is deprecated.

File: core/src/main/java/com/scalar/db/util/ProtoUtils.java [hunk 3]
```diff
@@ -74,18 +73,95 @@ public static com.scalar.db.rpc.Get toGet(Get get) {
...
   private static Key toKey(com.scalar.db.rpc.Key key, TableMetadata metadata) {
     Key.Builder builder = Key.newBuilder();
-    key.getValueList().forEach(v -> builder.add(toColumn(v.getName(), v, metadata)));
+
+    // For backward compatibility
+    if (!key.getValueList().isEmpty()) {
+      key.getValueList().forEach(v -> builder.add(toColumn(v.getName(), v)));
+    } else {
+      key.getColumnList().forEach(c -> builder.add(toColumn(c, metadata)));
+    }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/ProtoUtils.java:82]: If `Value` in ProtoBuf is used in a request, that indicates an old client sends the request. We need to handle this situation for backward compatibility.

File: core/src/main/java/com/scalar/db/util/ProtoUtils.java [hunk 4]
```diff
@@ -296,7 +358,14 @@ public static Mutation toMutation(com.scalar.db.rpc.Mutation mutation, TableMeta
...
     if (mutation.getType() == com.scalar.db.rpc.Mutation.Type.PUT) {
       Put put = new Put(partitionKey, clusteringKey);
-      mutation.getValueList().forEach(v -> put.withValue(toColumn(v.getName(), v, metadata)));
+
+      // For backward compatibility
+      if (!mutation.getValueList().isEmpty()) {
+        mutation.getValueList().forEach(v -> put.withValue(toColumn(v.getName(), v)));
+      } else {
+        mutation.getColumnList().forEach(c -> put.withValue(toColumn(c, metadata)));
+      }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/ProtoUtils.java:367]: Ditto.

File: core/src/main/java/com/scalar/db/util/ProtoUtils.java [hunk 5]
```diff
@@ -385,16 +454,21 @@ private static com.scalar.db.rpc.MutateCondition toCondition(MutationCondition c
...
-        toOperator(expression.getOperator()));
+    // For backward compatibility
+    if (expression.hasValue()) {
+      return ConditionBuilder.buildConditionalExpression(
+          toColumn(expression.getName(), expression.getValue()),
+          toOperator(expression.getOperator()));
+    } else {
+      return ConditionBuilder.buildConditionalExpression(
+          toColumn(expression.getColumn(), metadata), toOperator(expression.getOperator()));
+    }
```
Comment (brfrn169) [inline core/src/main/java/com/scalar/db/util/ProtoUtils.java:465]: Ditto.

File: server/src/main/java/com/scalar/db/server/DistributedStorageService.java [hunk 6]
```diff
@@ -69,8 +70,16 @@ public void get(GetRequest request, StreamObserver<GetResponse> responseObserver
...
+
           GetResponse.Builder builder = GetResponse.newBuilder();
-          result.ifPresent(r -> builder.setResult(ProtoUtils.toResult(r)));
+
+          // For backward compatibility
+          if (ProtoUtils.isRequestFromOldClient(request.getGet())) {
+            result.ifPresent(r -> builder.setResult(ProtoUtils.toResultWithValue(r)));
+          } else {
+            result.ifPresent(r -> builder.setResult(ProtoUtils.toResult(r)));
+          }
```
Comment (brfrn169) [inline server/src/main/java/com/scalar/db/server/DistributedStorageService.java:81]: If the request is from an old client, we need to return `Result` with the deprecated `Value` for backward compatibility.

File: server/src/main/java/com/scalar/db/server/DistributedStorageService.java [hunk 7]
```diff
@@ -268,8 +278,16 @@ private ScanResponse next(ScanRequest request) {
...
+
               ScanResponse.Builder builder = ScanResponse.newBuilder();
-              results.forEach(r -> builder.addResult(ProtoUtils.toResult(r)));
+
+              // For backward compatibility
+              if (ProtoUtils.isRequestFromOldClient(request.getScan())) {
+                results.forEach(r -> builder.addResult(ProtoUtils.toResultWithValue(r)));
+              } else {
+                results.forEach(r -> builder.addResult(ProtoUtils.toResult(r)));
+              }
```
Comment (brfrn169) [inline server/src/main/java/com/scalar/db/server/DistributedStorageService.java:289]: Ditto.

File: server/src/main/java/com/scalar/db/server/DistributedTransactionService.java [hunk 8]
```diff
@@ -292,8 +294,16 @@ private void get(GetRequest request, TransactionResponse.Builder responseBuilder
...
+
             GetResponse.Builder builder = GetResponse.newBuilder();
-            result.ifPresent(r -> builder.setResult(ProtoUtils.toResult(r)));
+
+            // For backward compatibility
+            if (ProtoUtils.isRequestFromOldClient(request.getGet())) {
+              result.ifPresent(r -> builder.setResult(ProtoUtils.toResultWithValue(r)));
+            } else {
+              result.ifPresent(r -> builder.setResult(ProtoUtils.toResult(r)));
+            }
```
Comment (brfrn169) [inline server/src/main/java/com/scalar/db/server/DistributedTransactionService.java:305]: Ditto.

File: server/src/main/java/com/scalar/db/server/DistributedTransactionService.java [hunk 9]
```diff
@@ -313,7 +323,14 @@ private void scan(ScanRequest request, TransactionResponse.Builder responseBuild
...
             List<Result> results = transaction.scan(scan);
             ScanResponse.Builder builder = ScanResponse.newBuilder();
-            results.forEach(r -> builder.addResult(ProtoUtils.toResult(r)));
+
+            // For backward compatibility
+            if (ProtoUtils.isRequestFromOldClient(request.getScan())) {
+              results.forEach(r -> builder.addResult(ProtoUtils.toResultWithValue(r)));
+            } else {
+              results.forEach(r -> builder.addResult(ProtoUtils.toResult(r)));
+            }
```
Comment (brfrn169) [inline server/src/main/java/com/scalar/db/server/DistributedTransactionService.java:332]: Ditto.

File: server/src/main/java/com/scalar/db/server/TwoPhaseCommitTransactionService.java [hunk 10]
```diff
@@ -336,7 +338,14 @@ private void get(
...
             Optional<Result> result = transaction.get(get);
             GetResponse.Builder builder = GetResponse.newBuilder();
-            result.ifPresent(r -> builder.setResult(ProtoUtils.toResult(r)));
+
+            // For backward compatibility
+            if (ProtoUtils.isRequestFromOldClient(request.getGet())) {
+              result.ifPresent(r -> builder.setResult(ProtoUtils.toResultWithValue(r)));
+            } else {
+              result.ifPresent(r -> builder.setResult(ProtoUtils.toResult(r)));
+            }
```
Comment (brfrn169) [inline server/src/main/java/com/scalar/db/server/TwoPhaseCommitTransactionService.java:347]: Ditto.

File: server/src/main/java/com/scalar/db/server/TwoPhaseCommitTransactionService.java [hunk 11]
```diff
@@ -357,7 +366,14 @@ private void scan(
...
             List<Result> results = transaction.scan(scan);
             ScanResponse.Builder builder = ScanResponse.newBuilder();
-            results.forEach(r -> builder.addResult(ProtoUtils.toResult(r)));
+
+            // For backward compatibility
+            if (ProtoUtils.isRequestFromOldClient(request.getScan())) {
+              results.forEach(r -> builder.addResult(ProtoUtils.toResultWithValue(r)));
+            } else {
+              results.forEach(r -> builder.addResult(ProtoUtils.toResult(r)));
+            }
```
Comment (brfrn169) [inline server/src/main/java/com/scalar/db/server/TwoPhaseCommitTransactionService.java:375]: Ditto.

**Review by feeblefakie — APPROVED** [review]
LGTM! Thank you!

**Review by Torch3333 — APPROVED** [review]
LGTM, thank you!

**Pull request merged**